### PR TITLE
Add missing non-method ops (typer layer)

### LINKED ISSUE #349: Make PredefinedFunc as single point of declaration
Same as `SMethod`(#344) but for predefined global functions.

### PR DESCRIPTION
**Pull request by greenhat**
#324 

Close #349 

Todo:
- [x] convert existing predefined global functions to use `PredefinedFunc`;
- [x] implement new predefined global functions (see spec);
- [x] fix ergo (pass network prefix to compiler);

### CODE REVIEW

File: src/main/scala/sigmastate/lang/SigmaPredef.scala [hunk 1]
```diff
@@ -15,16 +15,72 @@ object SigmaPredef {
     declaration: Lambda,
     /** Builder of SigmaIR node which is equivalent to function application
       * Rule: Apply(f, args) -->  irBuilder(f, args) */
-    irBuilder: (SValue, Seq[SValue]) => SValue
-  )
+    irBuilder: PartialFunction[(SValue, Seq[SValue]), SValue]) {
```
Comment (greenhat) [inline src/main/scala/sigmastate/lang/SigmaPredef.scala]: @ergomorphic What do you think of making `irBuilder` a `PartialFunction`? That makes function declaration succinct and much more readable with matching for arguments.

File: src/main/scala/sigmastate/lang/SigmaTyper.scala [hunk 2]
```diff
@@ -23,7 +23,7 @@ class SigmaTyper(val builder: SigmaBuilder) {
 
   private val tT = STypeIdent("T") // to be used in typing rules
 
-  private val predefFuncRegistry = new PredefinedFuncRegistry(builder)
+  private implicit val predefFuncRegistry: PredefinedFuncRegistry = new PredefinedFuncRegistry(builder)
```
Comment (ergomorphic) [inline src/main/scala/sigmastate/lang/SigmaTyper.scala]: @greenhat I think it is better to be a parameter of SigmaTyper. Next to SigmaBuilder.

File: src/main/scala/sigmastate/serialization/OpCodes.scala [hunk 3]
```diff
@@ -109,7 +109,7 @@ object OpCodes extends ValueCodes {
   val SliceCode            : OpCode = (LastConstantCode + 68).toByte
   val FilterCode           : OpCode = (LastConstantCode + 69).toByte
   val TreeLookupCode       : OpCode = (LastConstantCode + 70).toByte
-  val StringConcatCode     : OpCode = (LastConstantCode + 71).toByte
+  //  FREE                 : OpCode = (LastConstantCode + 71).toByte
```
Comment (ergomorphic) [inline src/main/scala/sigmastate/serialization/OpCodes.scala]: @greenhat why not to change TreeModificationsCode?

Comment (greenhat) [reply src/main/scala/sigmastate/serialization/OpCodes.scala]: @ergomorphic I've prematurely switched into "don't break compatibility" mode. Which does not make any sense at least until probably Testnet2. :) Fixed.

File: src/main/scala/sigmastate/lang/SigmaTyper.scala (hunk 2 above)
Comment (greenhat) [reply src/main/scala/sigmastate/lang/SigmaTyper.scala]: @ergomorphic Good idea! Thanks. Made it for binder as well.

File: src/main/scala/sigmastate/lang/SigmaPredef.scala (hunk 1 above)
Comment (ergomorphic) [reply src/main/scala/sigmastate/lang/SigmaPredef.scala]: @greenhat semantically it is not partial. For code readability you can use some helper which will transform PartialFunction to Function with error in default case.

Comment (ergomorphic) [reply src/main/scala/sigmastate/lang/SigmaPredef.scala]: @greenhat Do you use this partial function for pattern matching?

File: src/test/scala/sigmastate/lang/SigmaParserTest.scala [hunk 4]
```diff
@@ -519,9 +519,9 @@ class SigmaParserTest extends PropSpec with PropertyChecks with Matchers with La
   }
 
   property("ZKProof") {
-    parse("ZKProof { condition }") shouldBe Apply(SigmaPredef.ZKProofSym, IndexedSeq(Ident("condition")))
+    parse("ZKProof { condition }") shouldBe Apply(Ident("ZKProof", SFunc(SSigmaProp, SBoolean)), IndexedSeq(Ident("condition")))
```
Comment (aslesarenko) [inline src/test/scala/sigmastate/lang/SigmaParserTest.scala]: @greenhat maybe refer to PredefinedFunc here?

File: src/test/scala/sigmastate/lang/SigmaParserTest.scala [hunk 5]
```diff
@@ -519,9 +519,9 @@ class SigmaParserTest extends PropSpec with PropertyChecks with Matchers with La
...
     parse("ZKProof { sigmaProp(HEIGHT > 1000) }") shouldBe
-      Apply(SigmaPredef.ZKProofSym,
+      Apply(Ident("ZKProof", SFunc(SSigmaProp, SBoolean)),
```
Comment (aslesarenko) [inline src/test/scala/sigmastate/lang/SigmaParserTest.scala]: @greenhat same as above

File: src/test/scala/sigmastate/lang/SigmaTyperTest.scala [hunk 6]
```diff
@@ -87,7 +90,7 @@ class SigmaTyperTest extends PropSpec with PropertyChecks with Matchers with Lan
   }
 
   property("predefined functions") {
-    typecheck(env, "allOf") shouldBe AllSym.tpe
+    typecheck(env, "allOf") shouldBe SFunc(SCollection[SBoolean.type], SBoolean)
```
Comment (aslesarenko) [inline src/test/scala/sigmastate/lang/SigmaTyperTest.scala]: @greenhat Use PredefinedFunc?

File: src/main/scala/sigmastate/lang/SigmaPredef.scala [hunk 7]
```diff
@@ -1,91 +1,288 @@
...
+    private val funcNameToIrBuilderMap: Map[String, IrBuilderFunc] =
+      funcs.filter(_.irBuilder != undefined)
+        .map(f => f.name -> f.irBuilder)
+        .toMap
+
+    def irBuilderForFunc(name: String): Option[IrBuilderFunc] = funcNameToIrBuilderMap.get(name)
+  }
+
+  object PredefinedFuncApply {
+    def unapply(apply: Apply)(implicit registry: PredefinedFuncRegistry): Option[SValue] = apply.func match {
```
[the remark this answers is not in the document]
Comment (greenhat) [reply src/main/scala/sigmastate/lang/SigmaPredef.scala]: Done.

File: src/test/scala/sigmastate/lang/SigmaTyperTest.scala (hunk 6 above)
Comment (greenhat) [reply src/test/scala/sigmastate/lang/SigmaTyperTest.scala]: Fixed.

File: src/test/scala/sigmastate/lang/SigmaParserTest.scala (hunk 4 above)
Comment (greenhat) [reply src/test/scala/sigmastate/lang/SigmaParserTest.scala]: Fixed. Plus in other places where `Ident` was used.

File: src/test/scala/sigmastate/lang/SigmaParserTest.scala (hunk 5 above)
Comment (greenhat) [reply src/test/scala/sigmastate/lang/SigmaParserTest.scala]: Fixed.

**Pull request merged**